src/ceph-volume/ceph_volume/devices/lvm/deactivate.py

```diff
@@ -0,0 +1,96 @@
+import argparse
+import logging
+import sys
+from textwrap import dedent
+from ceph_volume import conf
+from ceph_volume.util import encryption, system
+from ceph_volume.api.lvm import get_lvs_by_tag
+
+logger = logging.getLogger(__name__)
+
+
+def deactivate_osd(osd_id=None, osd_uuid=None):
+
+    lvs = []
+    if osd_uuid:
+        lvs = get_lvs_by_tag('ceph.osd_fsid={}'.format(osd_uuid))
+    else:
+        lvs = get_lvs_by_tag('ceph.osd_id={}'.format(osd_id))
+
+    try:
+        data_lv = next(lv for lv in lvs if lv.tags['ceph.type'] in ['data', 'block'])
+    except StopIteration:
+        logger.error('No data or block LV found for OSD {}'.format(osd_id))
+        sys.exit(1)
+    conf.cluster = data_lv.tags['ceph.cluster_name']
+    logger.debug('Found cluster name {}'.format(conf.cluster))
+
+    tmpfs_path = '/var/lib/ceph/osd/{}-{}'.format(conf.cluster, osd_id)
+    logger.info('Unmounting osd-{}\'s tmpfs path at {}'.format(
+        osd_id,
+        tmpfs_path))
+    system.unmount(tmpfs_path)
+
+    for lv in lvs:
+        if lv.tags.get('ceph.encrypted', '0') == '1':
+            encryption.dmcrypt_close(lv.lv_uuid)
+        lv.deactivate()
```

Comment:
My (limited) understanding is that the LVM systemd units basically automagically activate all (non-encrypted) LVs on their own.  Should we really deactivate those?  It seems like it would mean activate later wouldn't work right, and/or that the system state would get out of sync with systemd.  My (uninformed) assumption was that we'd only deactivate the dmcrypt LVs (which is probably done by dmcrypt_close)?

Reply:
Yeah as I noted in the todo list: with the current state the `lvm activate` cmd needs to activate the lvs.
I understood we want both closing crypt devices and deactivate lvs/vgs. The latter is needed for containerized cluster (maybe on k8s?) so that an lv is released on container shutdown and can be attached to a new container.

Reply:
I'm not sure deactivating the LVs/VGs is addressed by this command, though.  Right now the issue @leseb has is related to EBS needing the LVs deactivaed to detach... it wouldn't be needed on bare metal IIUC.  And I seem to recall someone saying that there isn't a reliable shutdown hook that could run the deactivate command when a pod is killed.  For now, the EBS issue is being addressed by avoiding LVM entirely.

I'd suggesting making the initial version of this just undo the activate steps (so, just tmpfs and dmcrypt).  A later change could make the activate and deactivate lvchange steps together based on a flag or something, if we decide that's something we need?

Reply:
IMHO, all of that de-activation only makes sense when running in containers since we have no idea how systemd and udev rules are set up on the host (most of the time or at least no control over it because all our actions are in containers). So by activating/de-activating, we are basically re-implementing the systemd/udev rules logic in c-v, which is fine but I'd do it for containers only.

Reply:
Well, I'd argue that in the general case (no containers, normal systemd units) we should also be unmounting the tmpfs and stopping the dmcrypt device when the OSD stops.  For example, what happens if you want to yank a dmcrypted disk? Today you have to clean up the dmcrypt LVs manually, or just leave them hanging around along with the pinned original device, polluting the kernel's block device namespace.

If we stick to the dmcrypt LV and tmpfs, then we're not overlapping at all with what systemd/udev are doing

Reply:
Yeah what Sage said. Bassically any of the actions fail if the OSD is still running. If not unmount and encrypt are certainly unproblematic.
I do think we should also activate and de-activate lvs and vgs, if just for proper tiding, but I'm happy to leave that for a later PR.

Reply:
One more facet to this discussion: I don't think I fully understand yet, however I'm not seeing ehy ceph-volume needs to interact (as in call) systemd at all. I think currently things look like this:
ceph-volume@ systemd unit -> calls lvm trigger -> calls lvm activate -> calls systemctl start osd. Of course unless --no-systemd is passed to activate.

Wouldn't it make more sense to push the systemd dependencies into the systemd units and keep ceph-volume out of the whole deal?

Reply:
as in, have a post-start item for the device unit that starts the osd? something like that?

Reply:
> One more facet to this discussion: I don't think I fully understand yet, however I'm not seeing why ceph-volume needs to interact (as in call) systemd at all. I think currently things look like this:
> ceph-volume@ systemd unit -> calls lvm trigger -> calls lvm activate -> calls systemctl start osd. Of course unless --no-systemd is passed to activate.
> 
> Wouldn't it make more sense to push the systemd dependencies into the systemd units and keep ceph-volume out of the whole deal?

I think it depends on what parts you would like to move out. `ceph-volume` needs to be called for sure. I would be very hesitant to change this right now since it has achieved a level of robustness that wasn't possible with ceph-disk